rascal-lsp/src/main/rascal/lang/rascal/lsp/Outline.rsc

```diff
@@ -31,38 +31,58 @@ import String;
 import ParseTree;
 import lang::rascal::\syntax::Rascal;
 import util::LanguageServer;
+import util::ErrorRecovery;
 
 list[DocumentSymbol] outlineRascalModule(start[Module] \mod) {
     m= \mod.top;
-    children = [];
-
-    top-down-break visit (m) {
-        case (Declaration) `<Tags _> <Visibility _> <Type t> <{Variable ","}+ vars>;`:
-            children += [symbol(clean("<v.name>"), variable(), v@\loc, detail="variable <t> <v>") | v <- vars];
-
-        case (Declaration) `<Tags _> <Visibility _> anno <Type t> <Type ot>@<Name name>;`:
-            children +=  [symbol(clean("<name>"), field(), t@\loc, detail="anno <t> <ot>")];
 
-        case (Declaration) `<Tags _> <Visibility _> alias <UserType u> = <Type al>;`:
-            children += [symbol(clean("<u.name>"), struct(), u@\loc, detail="<u> = <al>")];
+    if (!(m has header) || hasErrors(m.header)) {
+        return [];
+    }
 
-        case (Declaration) `<Tags _> <Visibility _> tag <Kind k> <Name name> on <{Type ","}+ ts>;`:
-            children += [symbol(clean("<name>"), \key(), name@\loc, detail="tag <k> <name> on <ts>")];
+    children = [];
 
-        case (Declaration) `<Tags _> <Visibility _> data <UserType u> <CommonKeywordParameters kws>;`: {
-            kwlist = [symbol(".<k.name>", property(), k@\loc, detail="<k.\type>") | kws is present, KeywordFormal k <- kws.keywordFormalList];
-            children += [symbol("<u.name>", struct(), u@\loc, detail="data <u> <kws>", children=kwlist)];
+    top-down-break visit (m) {
+        case decl: (Declaration) `<Tags _> <Visibility _> <Type t> <{Variable ","}+ vars>;`:
+            if (!hasErrors(decl)) {
```

Comment:
I'm missing something here. If a node like this matches then it has enough structure to produce the symbol. What does it matter if there are errors inside?

Reply:
The point of the error trees is that they will be there but won't match with any of the expected language constructs. So you can ignore then typically.

For the case where you encounter one during recursion, a match on the error tree has to be added that ignores it or does something useful like a short yield: `"<error tree>"[..25]`

Reply:
Yield of an error tree is guaranteed to work: "<myStat>" will yield the entire string that was accepted to fill in for a real statement.